File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_price = prices[0];
        let mut max_profit: i32= 0;

        for i in 0..prices.len() {
            if min_price > prices[i] {
                min_price = prices[i];
            }

            if max_profit < (prices[i] - min_price) {
                max_profit = prices[i] - min_price;
            }
        }

        return max_profit
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = inf
        max_profit = 0

        for price in prices:
            # min price we've seen
            min_price = min(min_price, price)

            # if we sell those min_price today
            max_profit = max(max_profit, price - min_price)
        
        return max_profit